### PR TITLE
Ports -CloudyRedo

### PR DESCRIPTION
**Pull request by CloudyLopez**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | a custom method i wrote was:  (in works controller) @spotlight = Work.order(cached_votes_total: :desc).limit(10) this is where i kept my logic for the the top ten media works. as per the acts_as_voting gem required, per the schema of 'cached_votes_total' and then :desc bc we want the highest rated ones to be on top, the 10 is the the top ten works. [For another example pls see the work.rb file.]
Describe how you approached testing that model method. What edge cases did you come up with? | an edge case i came up with is in the works controller test, is asking it to give me a flash notice when finding a non existent work. 
What are `session` and `flash`? What is the difference between them? | session stays for the duration of the http request whereas flash 'flashes' the response once and it goes away. 
What was one thing that you gained more clarity on through this assignment? | lots of things, more clarity on the functions of all the billions of files but i did learn how to spot the blind spots with the tests so that was cool
What is the Heroku URL of your deployed application? | if it decides to work: https://cloudyschoice.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? | nope, this has been great.

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | I was hoping for more descriptive commit messages, and more granular commits
Comprehension questions	| For question No. 1, you mentioned something you did in your controller. However, the question asked for custom model methods you wrote... Also, your definitions of how flash and session aren't correct; come see me at some point to go over this material
**General** |
Rails fundamentals (RESTful routing, use of named paths) | x
Views are well-organized (DRY, use of semantic HTML, use of partials) | x, there's opportunity to refactor to use more partials, such as on the `works#index` view
Errors are reported to the user | some
Business logic lives in the models | Hm... there's some work here... but it's not quite what I expected. Also, there are a couple of pieces of logic I was looking for that aren't there
Models are thoroughly tested, including relations, validations and any custom logic | non-existent
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | UsersController: tested positive case of upvoting, and positive `/users` case, missing all others. WorksController: tested index and show, and no other actions
**Wave 1 - Media** |
Splash page shows the three media categories | no splash page... on `works#index` page, it does
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | cannot log out
The ID of the current user is stored in the session | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count |
Splash page contains a media spotlight | not correct
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | not quite correct... almost there
Individual user pages and the user list are present | x
**Optional - Styling** |
Bootstrap is used appropriately | x
Look and feel is similar to the original |
**Overall** |

Hi Cloudy!

Good work on this project. In general, you hit a lot of the learning goals:
- CRUD on `Work`s
- logging in using `session`
- behavior around logging in, such as not being able to vote until logging in, and not being able to vote twice

The following parts of this project submission give me concern:
- doing any custom model logic is absent
- the consistency of the wave completion-- there are parts of early waves that are incomplete, and were left incomplete before moving on
  - Why is logging out not a button on the page?
  - Why is sorting by vote not accomplished? Why is the spotlight feature pretty much there, but not accurate? (You list the top ten works sorted by vote, instead of just the top work sorted by vote)
  - Why is the splash page missing?
- the tests on controllers isn't up to expectations
- the tests on models isn't up to expectations. there are fixtures defined that don't get used
- the completion of business logic to demonstrate problem-solving in programming isn't there. A lot of this would have been accomplished in wave 2 with the custom model logic

And parts of this project submission leave me feeling ambiguous:
- The `acts_as_votable` gem sidesteps a lot of the learning goals of the project, and adds different kinds of complexities. In general, this isn't a problem (just moreso a funny thing), but it adds complexity in my ability to see how your web development skills are doing. It adds different kinds of complexities to your code base, too. Also...
- There's a lot of code in here that is identical to Kate's, as well as another student in C11's...

In general, I see some improvements in this project submission compared to your first one, and I see a lot of places in which there is still some improvements to go. Let me know if you have any questions

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+  has_many :votes
+  acts_as_voter
+  validates :name, presence: true, exclusion: { in: ['', nil] }
```
Comment (tildeee) [inline app/models/user.rb:4]: What does the `exclusion: { ... }` part do? Why do we need it or want it? How is it difference from `presence: true`?

**Comment by CloudyLopez** [comment]
Hi Dee,

If you have a moment, I would be happy to go over this with you.

Cheers,
Cloudy

File: app/models/vote.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates :user_id, presence: true
+  validates :work_id, presence: true
+  validates :value, presence: true, inclusion: { in: [1, -1], message: '%{value} is not a valid vote' }
```
Comment (tildeee) [inline app/models/vote.rb:9]: What does this line do? Is this part of the `acts_as_votable` gem?

File: app/models/vote.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+  def user
+    User.find(user_id)
+  end
```
Comment (tildeee) [inline app/models/vote.rb:13]: Hm... Not quite right... Why do we need an instance method named `user`? What does it mean to do `User.find(user_id)`? Where is `user_id` defined? When do we call this method?

File: app/models/work.rb [hunk 4]
```diff
@@ -1,3 +1,16 @@
...
+
+  def votes
+    votes = Vote.where('votable_id = ?', id)
+    puts 'vote.all.count', Vote.all.count
+    votes
+  end
 
+  def self.upvote(id)
+    puts id.to_s
+  end
```
Comment (tildeee) [inline app/models/work.rb:15]: Hm... I'm not sure what this method is... is this part of the `act_as_votable` gem? That being said, would we want it to just `puts id.to_s`? I don't know too many reasons why a `puts` statement would be valuable to the Rails app as a whole

File: config/routes.rb [hunk 5]
```diff
@@ -1,10 +1,11 @@
...
-  root to: "works#index"
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
-# root to: 'works#path'
-resources :works
-
+# frozen_string_literal: true
 
+Rails.application.routes.draw do
+  resources :works
+  resources :users, only: %i[index show]
```
Comment (tildeee) [inline config/routes.rb:5]: Interesting syntax... This is fine, but pretty inconsistent with the rest of the file and pretty difficult to read in this case; it may make more sense to just simply put `[:index, :show]`
